docs/en/middleware/supplemental.md

```diff
@@ -0,0 +1,521 @@
+
+**Note:** Only a single diff function can be registered for a given property during the lifetime of a composing widget or middleware, after which subsequent calls will be ignored. By default the rendering engine uses an algorithm that shallowly diffs objects and arrays, ignores functions, and equality checks all other property types. Setting a custom diff function overrides Dojo's default difference detection strategy for the property.
+
+**API:**
+
+```ts
+import diffProperty from '@dojo/framework/core/vdom';
+```
+
+-   `diffProperty(propertyName: string, diff: (current: any, next: any) => void)`
+    -   Registers the specified `diff` function that is called to determine if any differences exist between the `current` and `next` values of the widget's `propertyName` property.
+
+### `destroy`
+
+Assigns a function that is called on widget destruction, allowing any required resource teardown to take place.
+
+**Note:** `destroy()` can only be called once for each composing widget or middleware, after which further calls will be ignored. For advanced scenarios that need to conditionally add handles for execution when a widget is removed, a single destroy function should be registered that can keep track of and iteratively destroy all necessary resources.
+
+**API:**
+
+```ts
+import destroy from '@dojo/framework/core/vdom';
+```
+
+-   `destroy(destroyFunction: () => void)`
+    -   Sets the `destroyFunction` that will be called when the current widget is destroyed. Setting a function will override any destroy function previously set for the widget.
+
+### `getRegistry`
+
+Provides access to the widget's own `Registry` instance, as well as the root application `Registry` if required, via a handler interface.
+
+**Note:** The registry is an advanced concept not typically required when writing Dojo applications. It is mainly used internally by the framework to implement more advanced user-facing functionality such as [Dojo stores](../stores/basic-usage.md).
+
+**API:**
+
+```ts
+import getRegistry from '@dojo/framework/core/vdom';
+```
+
+-   `getRegistry(): RegistryHandler | null`
+    -   Returns the `RegistryHandler` for the current widget, or `null` if the widget has not yet been fully initialized.
+
+## Optional middleware
+
+Dojo provides a variety of optional middleware that widgets can include when needing to implement specific requirements.
+
+### `block`
+
+Allows widgets to execute modules known as **blocks** within Node.js at build time. Typically used as part of build-time rendering.
+
+Described in detail in the [Building](../building/supplemental.md#build-time-rendering) reference guide.
+
+**API:**
+
+```ts
+import block from '@dojo/framework/core/middleware/block';
+```
+
+-   `block<T extends (...args: any[]) => any>(module: T)`
+    -   Executes the specified block module and returns its result.
+
+### `breakpoint`
```

Comment:
This should probably go below `resize`, it looks like they are ordered alphabetically at the moment but I think they should be grouped together by functionality and ordered by "importance"/"how often they would be used" by consumers.

Reply:
Off the top of my head the order should be something like:

* cache
* icache
* theme
* i18n
* dimensions
* intersection
* resize
* breakpoint
* store
* focus
* injector
* block

Reply:
Done - I've also moved the core middleware list after the optional set so the entire set is ordered by usage frequency